crypto/evp/evp_enc.c

```diff
@@ -589,6 +593,9 @@ int EVP_EncryptFinal_ex(EVP_CIPHER_CTX *ctx, unsigned char *out, int *outl)
     size_t soutl;
     int blocksize;
 
+    if (outl != NULL)
```

Comment:
`outl` does not seem to be documented as optional...so why change it to be so?

Reply:
Yeah, so when I went to write "*outl = 0" the sirens went off in my head that it might not be safe to de-reference that pointer. It turns out that the non-legacy path is already doing this (I didn't check further down the legacy call stack yet), so I tried to add safety there as well.

As I see it, there's two options for improving how a NULL outl is handled within the context of these functions. Obviously no existing client code is going to be relying on the existing behavior of de-referencing it, so I think we can choose freely without compatibility concern. The first is to explicitly return an error. The second is as I've done. The reason that I chose this is that said output parameter may not always be useful to the caller. Depending on what the caller knows about the block alignment of the input data and so on, they may actually know a priori what outl should be set to in the success case. (At minimum the API requires them to know the upper bound on it to safely size the out buffer.)

I don't have a particularly strong objection to switching over to the first approach, although I would note that the legacy path's behavior should be evaluated in that case.

Reply:
FWIW, it looks like the legacy paths also de-reference without checking.

Reply:
There are 2 schools of though wrt to NULL checks on mandatory parameters:
1) We never NULL check. If an application doesn't supply the parameter then that's an application error - so a crash is the fault of the application, e.g. if you pass NULL to various libc functions such as memcpy you can expect a crash, so why would an application expect something else from OpenSSL.
OR
2) We should always NULL check and return an error immediately if necessary. This prevents library crashes when perhaps only one thread in the application behaved badly (why bring the whole server down in that case)

For a long time (1) was the predominant thinking and it is common practice within the codebase to omit NULL checks. More recently (2) has had its proponents. It's never been resolved.

So, IMO in would be acceptable to either have no NULL check, or have a NULL check and immediately error.

What we have here is a change in the behaviour to make this argument optional. I don't particularly see a strong need for that but wouldn't necessarily object if there was support for it. However I don't think such a behaviour change should be done as part of some other PR with a different objective (i.e. this one). If we want to change the behaviour in that way it should have a PR all of its own, along with associated documentation updates.

Reply:
Fair enough, I'll just convert this to return an error in an "else" branch to the "if" you're noting.

FWIW, I'm not sure that I'd ALWAYS put myself in camp 2, but certainly with respect to these APIs I would. That's based on the fact that they could conceivably be used sensibly by a client without subsequently reading outl. How a function is formally documented certainly does matter, but I think that assuming a programmer has always read the documentation isn't a great idea. If you take a general argument of "it's documented that you're supposed to do X, if the client doesn't do that then it's on them" to its logical conclusion you never need to return any kind of "bad input" errors. So I think the premise that at least SOME documented requirements should be explicitly checked within the implementation is generally an accepted one, it's just a question of where one draws the line. For my money, the question is really how obvious it is from only the method signature that a parameter is "mandatory."